src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -528,18 +568,60 @@ def getMSOutputTransferInfo(self, wflow):
             wflow['TransferDone'] = True
         return wflow
 
-    def getRucioRules(self, wflow, gran, rucioAcct):
+    def setParentDatasets(self, wflow):
+        """
+        Used to resolve parent datasets for a workflow.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :return:         The workflow object
+        """
+        if wflow['InputDataset'] and wflow['IncludeParents']:
+            childDataset = wflow['InputDataset']
+            parentDataset = findParent([childDataset], self.msConfig['dbsUrl'])
+            if parentDataset:
+                wflow['ParentDataset'].append(parentDataset[childDataset])
```

Comment:
This line seems unnecessary.

Reply:
No, it is not unnecessary

Reply:
My bad, I just noticed the line below is again on the `ParentDataset` key.

However, looking at the findParent implementation, the line below (making unique elements) is not really needed, since it returns a single parent dataset for every children.
By the way, there is a flaw here because it does not account for an exception in the parentage resolution (value = None).

Reply:
I knew which line you were talking about. Again, it is needed, and not because the `findParents()` returns more than one dataset, but because we may call it twice, and we do not want to accumulate in the 'ParentDataset' list the same dataset which will show up twice.

Reply:
Hi @amaltaro , About the returned `value=None`, I just made a commit that should address this.

Reply:
I had to force push one more commit in adjacent to the above mentioned. I was not handling the exception I was raising and the standard long error message from WMCoreException was flooding the logs.

Reply:
Just to make sure I understand your point, are you saying that we can call this method `setParentDatasets` multiple times for the same workflow within the same cycle? That's the only explanation to have a list here.

The `findParent` function returns a string as value:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/Unified/Common.py#L651

Reply:
Yes, this i exactly what I am saying. 

>The findParent function returns a string as value:

`findParent` returns either a dictionary with keys names set as the child dataset name and values set as the primary dataset (the value could eventually be None if the child dataset has no parent discovered), or `None` instead of a dictionary in the cases when the DBS service fails to answer the query. I hope I am not missing something here...

Reply:
I think we agree on that. Being explicit, findParent() is meant to return data formatted as:
```
{"child_dataset_nameA": "parent_dataset_name",
 "child_dataset_nameB": "parent_dataset_nameX", etc}
```

with such data, these 2 lines you have in the current code:
```
wflow['ParentDataset'].append(parentDataset[childDataset])
wflow['ParentDataset'] = list(set(wflow['ParentDataset']))
```

could be converted to one line:
```
wflow['ParentDataset'] =[parentDataset[childDataset]]
```

Your comment also made me realize a mistake on the way the code parses errors/exceptions in the findParent() function. Shouldn't this if statement:
```
if parentDataset is None:
```

be actually converted to:
```
if not parentDataset.get(childDataset):
```
?

Reply:
> `wflow['ParentDataset'] = [parentDataset[childDataset]]`

Agreed

> `if parentDataset is None:`
This line I wrote because like that , because I wanted to threat separately the cases when `findParent` returns just a   `None` value instead of the normal dictionary, which is a sign for a DBS service error, but not a workflow misconfiguration or a WMCore error. In those cases we should skip this workflow and postpone its treatment/cleaning until next poling cycle.

Reply:
Done